sycl/doc/design/OffloadDesign.md

```diff
@@ -0,0 +1,282 @@
+implied options set via target options such as `-fsycl-targets=intel_gpu_skl`
+will be processed by a new options to the wrapper, `--gen-tool-arg=<arg>`
+
+#### spir64_fpga support
+
+Compilation behaviors involving AOT for FPGA involve an additional call to
+the either `aoc` (for Hardware) or `opencl-aot` (for Simulation).  This call
+occurs after the post-link step performed by sycl-post-link.  Additional
+options passed by the user via the `-Xsycl-target-backend=spir64_fpga <opts>`
+command will be processed by a new options to the wrapper,
+`--fpga-tool-arg=<arg>`
+
+The FPGA target also has support for additional generated binaries that
+contain intermediate files specific for FPGA.  These binaries (aoco, aocr and
+aocx) can reside in archives and are treated differently than traditional
+device binaries.
+
+When using the `-fintelfpga` option to enable AOT for FPGA, there are
+additional expectations during the compilation.  Use of the option will enable
+debug generation and also generate dependency information.  The dependency
+generation should be packaged along with the device binary for use during
+the link phase.  If we are not generating an object and performing compilation
+directly to the final executable, a new option named `--fpga-tool-deps=<arg>`
+will be needed to pass along the name of the dependency files created during
+the compile.  The dependency information is only used when compiling for
+hardware.
+
+#### spir64_x86_64 support
+
+Compilation behaviors involving AOT for CPU involve an additional call to
+`opencl-aot`.  This call occurs after the post-link step performed by
+`sycl-post-link`.  Additional options passed by the user via the
+`-Xsycl-target-backend=spir64_x86_64 <opts>` command will be processed by a new
+option to the wrapper, `--cpu-tool-arg=<arg>`
+
+### Integration of the sycl-aspect-filter
+
+### Wrapping of device image
+
+Once the device binary is pulled out of the fat binary, the binary must be
+wrapped and provided the needed entry points to be used during execution.  This
+is performed by the during the link phase and controlled by the
+`clang-linker-wrapper`.
+
+It is expected that the wrap information that is generated to be wrapped
+around the device binary will match current wrapping information that is used
+for the exiting offload model.  The wrapping in the old model is using the
+`clang-offload-wrapper` tool.
+
+### Integration of llvm-foreach
+
+Use of `llvm-foreach` is used frequently during the offloading process.  The
+functionality will persist with the new model.  Steps using `llvm-foreach`
+have the capability of kicking of the steps in a parallel manner by using the
+`-fsycl-max-parallel-link-jobs` option on the command line.  Information
+about the parallel jobs will need to be passed through the
+`clang-linker-wrapper` to be properly processed.  The option will be named
+`--parallel-link-sycl=<arg>` to be consumed and used during `llvm-foreach`
+toolchain events.
```

Comment:
We should aim for that in the future, but there could be a caveat: we don't know if all of those tools support multi-threaded environment. For example: the SPIR-V translator has a library API, but I doubt that it provides any guarantees that those APIs can be called in parallel. It could be some hidden global states which may caused problems or some other issues.

I'm not saying that we shouldn't aim for inlining as much as possible into `clang-linker-wrapper` to reduce overheads on I/O and spawning separate processes, but there could be some obstacles on that path with some of the tools.

Reply:
Where do the threads come from?  Is `clang-linker-wrapper` currently multi-threaded in its implementation?

Reply:
> Is `clang-linker-wrapper` currently multi-threaded in its implementation?

I don't know for sure, but it has `wrapper-jobs` option which implies some parallelization - I guess it happens somewhere in thinLTO.

> Where do the threads come from?

My assumption that if we want to both _inline_ external tools into `clang-linker-wrapper` and _launch them in parallel_, that will require threads and not separate processes.